File: gemgen.hs
gems        = zip "abcdefg" [1, 4, 1, 4, 2, 1, 3]

generate prefix gems len =
  let
    letters = gems >>= \(c, n) -> if n > 0 then [c] else []
    gems' c gs =
      gs >>= \(c', n) -> if c == c' then [(c', n - 1)] else [(c', n)]
  in
   if len > 0
   then
     prefix : (letters >>= \c ->
                generate (prefix ++ [c]) (gems' c gems) (len-1))
   else
     [prefix]

xs = zip [0..] (generate "" gems 16)

lookup_string goal ((i,s):xs) = 
  if s == goal
  then
    putStrLn $ (show i) ++ ":" ++  s
  else
    lookup_string goal xs

main = lookup_string "eagcdfbe" xs

File: gemgen.rb
#!/usr/bin/env ruby

class GemGen
  def initialize(len, prefix, gems)
    @len = len
    @prefix = prefix
    @gems = gems
  end

  def i2c(i)
    "abcdefg"[i]
  end

  def each &b
    b.call(@prefix)

    if @len > 0
      @gems.each_with_index do |n, i|
        if n > 0
          subgems = @gems.dup
          subgems[i] -= 1
          c = i2c(i)
          newprefix = "#{@prefix}#{c}"
          g = GemGen.new(@len - 1, newprefix, subgems)
          g.each {|str| b.call(str)}
        end
      end
    end
  end
end

Enumerator.new(GemGen.new(16, "", [1, 4, 1, 4, 2, 1, 3])).each_with_index do |s, i|
  if s == "eagcdfbe"
    puts "#{i}:#{s}"
    break
  end
end